### PR TITLE
Use OpenTok signaling In place of PubNub for Archive and Connection management

### PR DESCRIPTION
**Pull request by normanargueta**

### CODE REVIEW

**Comment by beHaze** [comment]
LGTM

**Comment by jeffswartz** [comment]
When I join the room, the archive status indicator should indicate the number of existing archives (after the client receives that info). It does not. You only see the number of existing archives once an archive is started after a client joins.

**Comment by jeffswartz** [comment]
Maybe we should change the name of web/js/models/firebase.js to web/js/models/archive.js and change FirebaseModel to `ArchiveModel`. @normanargueta, wdyt?

File: server/serverMethods.js [hunk 1]
```diff
@@ -447,6 +442,36 @@ function ServerMethods(aLogLevel, aModules) {
...
+        if (enableArchiveManager) {
+          var fbArchivesCallback = function (archives) {
+            if (archives.val()) {
+              tbConfig.otInstance.signal(
+                usableSessionInfo.sessionId,
+                null,
+                {
+                  type: 'archives',
+                  data: JSON.stringify({
+                    _head: { // TODO: use the same method that's implemented in client to wrap signal msg
```
[the remark this answers is not in the document]
Comment (jeffswartz) [reply server/serverMethods.js]: Can we just change this to `data: JSON.stringify(archives.val())`? What is the point of sending the `_head` property, which has static data?

File: server/serverMethods.js [hunk 2]
```diff
@@ -447,6 +442,36 @@ function ServerMethods(aLogLevel, aModules) {
...
+                    },
+                    data: archives.val()
+                  })
+                },
+                function (error) {
+                  if (error) {
+                    return logger.log('Get archives error:', error);
+                  }
+                },
+              );
```
[the remark this answers is not in the document]
Comment (jeffswartz) [reply server/serverMethods.js]: Maybe we should add a `lint` task to the package.json file.

File: server/serverMethods.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (normanargueta) [reply server/serverMethods.js]: I'm still thinking what's the best approach to solve that. The problem is the web client parses all the incoming signal messages with a custom method: https://github.com/opentok/opentok-rtc/blob/develop/web/js/helpers/OTHelper.js#L122 and that method works if the signal message has been wrapped with this method: https://github.com/opentok/opentok-rtc/blob/develop/web/js/helpers/OTHelper.js#L69

[the remark this answers is not in the document]
Comment (jeffswartz) [reply server/serverMethods.js]: I wasn't aware of that `receiveMultipartMsg()` method in OTHelper. OK, then it makes sense to leave in the `_head` property.

File: server/firebaseArchives.js [hunk 3]
```diff
@@ -139,6 +118,22 @@ function FirebaseArchives(aRootURL, aSecret, aCleanupTime, aLogLevel, pubnubSubK
             });
         });
       },
+      subscribeArchiveUpdates(sessionId, sendSignalCallback) {
+        return new Promise((resolve) => {
```
[the remark this answers is not in the document]
Comment (normanargueta) [reply server/firebaseArchives.js]: Fixed!

File: server/firebaseArchives.js [hunk 4]
```diff
@@ -139,6 +118,22 @@ function FirebaseArchives(aRootURL, aSecret, aCleanupTime, aLogLevel, pubnubSubK
...
+          fbRootRef.child(sessionId + '/archives').on('value', sendSignalCallback);
+        });
+      },
+      storeConnection(connection, sessionId) {
+        return new Promise((resolve) => {
```
[the remark this answers is not in the document]
Comment (normanargueta) [reply server/firebaseArchives.js]: Fixed!

File: web/js/helpers/requests.js [hunk 5]
```diff
@@ -92,13 +92,29 @@
     return sendXHR('DELETE', server + '/archive/' + id);
   }
 
+  function saveConnectionFirebase(connection, sessionId) {
```
[the remark this answers is not in the document]
Comment (normanargueta) [reply web/js/helpers/requests.js]: Sounds good for me, I renamed the method

File: web/js/helpers/requests.js [hunk 6]
```diff
@@ -92,13 +92,29 @@
...
+    return sendXHR('POST', server + '/storeConnection/', JSON.stringify({
+      connection: connection,
+      sessionId: sessionId
+    }), 'application/json');
+  }
+
+  function deleteConnectionFirebase(connection, sessionId) {
```
[the remark this answers is not in the document]
Comment (normanargueta) [reply web/js/helpers/requests.js]: Sounds good for me, I renamed the method

File: views/room.ejs [hunk 7]
```diff
@@ -17,7 +17,7 @@
     <link rel="stylesheet" href="/css/room.opentok.css">
     <link rel="stylesheet" href="/css/buttons.css" />
 
-    <% if (useGoogleFonts) { %>
+    <% if (useGoogleApi) { %>
```
[the remark this answers is not in the document]
Comment (normanargueta) [reply views/room.ejs]: Fixed!

**Comment by normanargueta** [comment]
@jeffswartz thanks for the complete review, I have solved some of the issues you pointed however I still working in some others yet. We are going to merge this PR and for the fixes pending I'm going to make another PR direct to develop branch

**Pull request merged**